From 89a44e440ce1b9c459b080f61d514659d3d12572 Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Tue, 4 Sep 2018 18:29:37 +0900
Subject: [PATCH 368/909] rcar-vin: rcar-csi2: Fix FLD register setting order

commit c4f2e8729024cd4fac3b44dafe7bef3f645e6062 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

According to latest H/W manual v1.00, FLD register setting
must be set before setting ICLK_NONSTOP bit in LINK_CNT.
This patch fixes it.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/rcar-vin/rcar-csi2.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/media/platform/rcar-vin/rcar-csi2.c b/drivers/media/platform/rcar-vin/rcar-csi2.c
index 1e320d48b779..304b8211675b 100644
--- a/drivers/media/platform/rcar-vin/rcar-csi2.c
+++ b/drivers/media/platform/rcar-vin/rcar-csi2.c
@@ -527,8 +527,6 @@ static int rcsi2_start(struct rcar_csi2 *priv, struct v4l2_subdev *nextsd)
 	rcsi2_write(priv, PHTC_REG, 0);
 
 	/* Configure */
-	rcsi2_write(priv, FLD_REG, fld | FLD_FLD_EN4 |
-		    FLD_FLD_EN3 | FLD_FLD_EN2 | FLD_FLD_EN);
 	rcsi2_write(priv, VCDT_REG, vcdt);
 	rcsi2_write(priv, VCDT2_REG, vcdt2);
 	/* Lanes are zero indexed. */
@@ -558,6 +556,13 @@ static int rcsi2_start(struct rcar_csi2 *priv, struct v4l2_subdev *nextsd)
 	rcsi2_write(priv, PHYCNT_REG, phycnt);
 	rcsi2_write(priv, LINKCNT_REG, LINKCNT_MONITOR_EN |
 		    LINKCNT_REG_MONI_PACT_EN | LINKCNT_ICLK_NONSTOP);
+
+	if (priv->mf.field != V4L2_FIELD_NONE)
+		rcsi2_write(priv, FLD_REG, fld | FLD_FLD_EN4 |
+			    FLD_FLD_EN3 | FLD_FLD_EN2 | FLD_FLD_EN);
+	else
+		rcsi2_write(priv, FLD_REG, 0);
+
 	rcsi2_write(priv, PHYCNT_REG, phycnt | PHYCNT_SHUTDOWNZ);
 	rcsi2_write(priv, PHYCNT_REG, phycnt | PHYCNT_SHUTDOWNZ | PHYCNT_RSTZ);
 
-- 
2.17.1

